### ISSUE TITLE
Support JPEG XL images

### ISSUE THREAD

**Issue by nekohayo** [libgd/libgd#699]
Why: JPEG XL is pretty much the future of all image formats. It is a new royalty-free image codec targeting the image quality as found on the web, providing about 60% size savings when compared to original JPEG at the same perceptual quality, while supporting modern features like HDR, animation, alpha channel, lossless JPEG recompression, lossless and progressive modes. It is based on Google's PIK and Cloudinary's FUIF, and is in the final steps of standardization with ISO.

Explanation here: https://www.slideshare.net/cloudinarymarketing/imagecon-2019-jon-sneyer

Reference implementation here: https://gitlab.com/wg1/jpeg-xl

Chromium/Chrome and Firefox's dev versions already have initial implementations, so I'm hopeful that this will start becoming available within months...

Imagemagick already supports jpegxl, but I think Wordpress would want gd to support it too before it can be considered over there: https://core.trac.wordpress.org/ticket/52788 and https://github.com/WordPress/performance/issues/12

**Comment by pierrejoye** [comment]
Sounds good to me. Not sure when we can do it, but it will definitively be added :)

If you know C, help welcome as I am focusing on the GD Draw 2D vector API (canvas-like APIs for GD with quality antialiasing)  :)

**Comment by morsssss** [comment]
I don't know if I'll get to this, but it's possible I could work on this in October! But perhaps someone else will beat me to it :)

**Comment by cmb69** [comment]
A new image format every other year – what's wrong with GIF? ;)

**Comment by morsssss** [comment]
Can we support the [Bartlane cable picture transmission system](https://en.wikipedia.org/wiki/Bartlane_cable_picture_transmission_system)?

**Comment by pierrejoye** [comment]
hehe Ben  :) it looks like a fun format.

This new jpeg version looks interesting, let see the adoption.  Maybe it
will work better than the previous attempts.

**Comment by sdevrieze** [comment]
https://caniuse.com/?search=jxl

**Comment by morsssss** [comment]
@pierrejoye I see your point. I thought this was going to go into Chromium soon, but maybe it's best to wait to see a browser or two supporting this in production.

**Comment by nekohayo** [comment]
Well I doubt these two browser vendors would back down from an already implemented support in their testing builds, wouldn't make sense to me that they would throw away this work, especially as big players in the industry (not just Google and Cloudinary but also Facebook Infrastructures, Adobe including one of the pioneers of the PDF format, etc.) have publicly and unequivocally voiced enthusiasm and support for this initiative (there's proof in the comments in the Firefox ticket). I'm reasonably confident JPEG XL adoption will happen, and that it's a matter of time before the browser implementations "in testing" trickle down to production? It's nicer if the other pieces of the industry puzzle like gd (which Wordpress seems to be waiting on) end up being ready about the same time, though I understand there is no guarantee here. Just stating my slight optimism :)

**Comment by pierrejoye** [comment]
oh it will happen :)

the question is only about the priority.  I am personally busy with 3.0 and
the new internal image formats, much better and high quality drawing api
(you can see the wiki and the surface branch). This is on top of my list .

**Comment by vapier** [comment]
Chrome/etc... have dropped things in the past when they don't pan out, and the JPEG IJG have produced duds before.  so "big names" are by no means a guarantee.

if you want to contribute a PR to implement the functionality, then by all means, go for it.

**Comment by cmb69** [comment]
Especially if this is about WordPress, I see no urgent need. AFAIK, WordPress does not yet generally support PHP 8.0, and support for new image formats will have to wait for PHP 8.2 anyway.

**Comment by adamsilverstein** [comment]
Absolutely agree this will take some time to make its way into WordPress (although we do support PHP 8.0 in WordPress now).

**Comment by prlbr** [comment]
Please support encoding progressive JXLs so that the content of an image can be displayed (in a lower quality) already before the file is fully downloaded on the web. This is highly important for a smooth user experience.

Note: GD already supports progressive JPEGs: https://libgd.github.io/manuals/2.3.0/files/gd-c.html#gdImageInterlace

**Comment by mstrelan** [comment]
Both Firefox and Chrome announced intent to ship support for JPEG XL:

- https://hacks.mozilla.org/2026/08/intent-to-ship-jpeg-xl/
- https://groups.google.com/a/chromium.org/g/blink-dev/c/-gDojQbDPRI

**Comment by pierrejoye** [comment]
oh, master has full JXL support already. Animations included.

that's a jxl anim converted to webm (ffmpeg) f.e., using jxl anim from the new 2D canvas api  (path transform).

https://github.com/user-attachments/assets/b82c0055-9adb-4c51-990a-de98dfe6e893

(original file):


[anim_jxl.zip](https://github.com/user-attachments/files/31405063/anim_jxl.zip)

**Comment by pierrejoye** [comment]
closing as solved :)

**Comment by pierrejoye** [comment]
> Please support encoding progressive JXLs so that the content of an image can be displayed (in a lower quality) already before the file is fully downloaded on the web. This is highly important for a smooth user experience.
> 
> Note: GD already supports progressive JPEGs: https://libgd.github.io/manuals/2.3.0/files/gd-c.html#gdImageInterlace

Please open a separate issue for progressive. That's something I skip.

It will be only for still JXL tho', adding this for anims defeat pretty much anything jxl does for anims and much slower. For still, it only slightly increases the file size. Where "slightly" can be quite a bit for some FP formats tho' (which are supported by gd now too).

**Comment by morsssss** [comment]
@mstrelan thanks for sticking this news here! I'd been told Chrome would never support it :)

**Comment by pierrejoye** [comment]
> [@mstrelan](https://github.com/mstrelan) thanks for sticking this news here! I'd been told Chrome would never support it :)

chromium based browsers already support JXL, still and animations, You can try with the file I posted earlier :)

**Comment by mstrelan** [comment]
> > Please support encoding progressive JXLs so that the content of an image can be displayed (in a lower quality) already before the file is fully downloaded on the web. This is highly important for a smooth user experience.
> > Note: GD already supports progressive JPEGs: https://libgd.github.io/manuals/2.3.0/files/gd-c.html#gdImageInterlace
> 
> Please open a separate issue for progressive. That's something I skip.
> 
> It will be only for still JXL tho', adding this for anims defeat pretty much anything jxl does for anims and much slower. For still, it only slightly increases the file size. Where "slightly" can be quite a bit for some FP formats tho' (which are supported by gd now too).

Opened https://github.com/libgd/libgd/issues/1032

**Comment by pierrejoye** [comment]
side not, I may see if I can support google's jxl-rs which will be used by firefox as well. I did not check yet but I hope the C interface is compatible. If not, it would be like a full new codec on libgd's side :)
